### DISCUSSION TITLE
Why is the URI required for the document from a string?

### DISCUSSION THREAD

**Discussion by Lemour-sudo** [eclipse-langium/langium#1395] [Q&A] [answered]
I would assume that a URI would not be necessary to specify when extracting a Langium document from a string in code instead of extracting from an external file.

However, this is not the case with the functions provided for extracting a document from a string in code: `fromString<T extends AstNode = AstNode>(text: string, uri: URI): LangiumDocument<T>;`

Is there any reason why the URI is required in this case?

**Comment by msujew** [comment, marked as answer]
Hey @Lemour-sudo,

in essence, the `uri` is the unique identifier of a document. Since Langium is mostly designed to support language servers, and all documents inside of a language server require a uri identifier (otherwise they're not addressable), we decided to make it a mandatory property. From a maintenance perspective, having it optional would make every bit of code dealing with those documents very verbose.
